### ISSUE TITLE
Performance regression in 4.1.5?

### ISSUE THREAD

**Issue by kelly-tock** [reduxjs/reselect#560]
Upgraded and application from 4.0.0 to 4.1.5. We are also using `re-reselect` as well as `defaultMemoize` in several places as a more general memoization solution.  A screen on our app is essentially unresponsive. 

it seems like memoization is broken, as if I observe profiles in chrome dev tools with 4.0.0 and 4.1.5, when I click on a thing for example, all the sub items in a grid of items re-render, but in 4.0.0 this is not the case. all jest tests involving selectors are passing. 

have there been any subtle pattern changes when folks have upgraded that are needed?  I see some changes involving `defaultMemoize` but it seems harmless. 

I have also tried removing `re-reselect` and using the new built in `maxSize` config options. 

any insight would be much appreciated!  i'm still working on narrowing down the issue, but  wanted to ask. Was very surprised that it all seems to work except on one screen on our application that is very data intensive.

**Comment by markerikson** [comment]
The internal implementation definitely changed, but all the existing tests are passing. Can you post an example project that reproduces the issue, and ideally one that shows both 4.0 and 4.1.5 at the same time?  (like, added as separate dependencies and then running the same logic back-to-back or something)

**Comment by kelly-tock** [comment]
yep, i'll try it with the selectors I have in mind.

**Comment by kelly-tock** [comment]
still trying to isolate the issue, as our app has 100s of selectors and lots of long long long selector chains of computing something from 4-5 selectors, then computing off of that as well. what is interesting to note that with the memory checkbox enabled in chrome dev tools profiling, i'm seeing this pattern:

4.1.5:
![image](https://user-images.githubusercontent.com/39061683/145594626-5308f8e2-733a-4322-ac29-bdd3122e7cf2.png)

which seems indicative of a memory leak. 

with 4.0.0 this same set of actions:
![image](https://user-images.githubusercontent.com/39061683/145598999-237639d6-3d74-43f3-aaff-22867c376d85.png)

i'm continuing to investigate, it may be some pattern we have in our codebase that's an older pattern that's discouraged now or something like that.

**Comment by markerikson** [comment]
Huh, that's _very_ surprising. I'm assuming you didn't configure any of these selectors for a larger cache size. There's specific handling internally for a cache size of 1, and I don't see anything immediately obvious that would be an issue:

https://github.com/reduxjs/reselect/blob/v4.1.5/src/defaultMemoize.ts#L21-L44

**Comment by kelly-tock** [comment]
nope, this is testing 4.1.5 as a drop in replacement. we have a few spots where we're using `re-reselect` for cache size stuff ATM.

**Comment by kelly-tock** [comment]
is there anything with `undefined` vs that `NOT_FOUND` value?

**Comment by markerikson** [comment]
Not sure what you're asking - clarify?

**Comment by kelly-tock** [comment]
I thought I saw in the various release notes something about returning undefined explicitly for a selector and a mention of this `NOT_FOUND` constant, but i'm just thinking out loud. we return undefined in selectors all over the place.

**Comment by markerikson** [comment]
Yeah, I had to switch to that `NOT_FOUND` sentinel value specifically to disambiguate with `undefined` so that selectors could store that as a result.

**Comment by kelly-tock** [comment]
so do we need to do something on our end to change to that value?

**Comment by markerikson** [comment]
No, that's purely an internal thing so that the logic above it can confirm that "the cache did not find a match for the key we're checking against".

**Comment by kelly-tock** [comment]
still isolating the issue, but from observing so far, a memory leak persisted, but got a little better every time I removed more selectors in our application from running. is there something internally that keeps track of all the selectors that is doing something to cause more memory consumption?

**Comment by markerikson** [comment]
Nope. Each selector is its own separate cache - there's nothing global.

It kinda sounds like it's less of a "memory leak" per se, and more just "caching".  It sounds as if the selectors are keeping more data references alive than they used to, and that's adding more.  (I realize there's a small difference in semantics there :)  )

**Comment by kelly-tock** [comment]
still haven't found it, but I thought maybe posting the relevant areas of 2 profiles of the app i am working on, one with latest, and one with 4.0.0.

4.1.5:
![image](https://user-images.githubusercontent.com/39061683/146586583-79290ab5-b0c8-49e3-8ec8-832a004e355e.png)
![image](https://user-images.githubusercontent.com/39061683/146586669-9bceb236-b8ea-4058-8243-a982eebe5758.png)


same area of the profile for 4.0.0:
![image](https://user-images.githubusercontent.com/39061683/146586837-8f632432-a1c4-4713-87bc-36710acf0e31.png)

**Comment by kelly-tock** [comment]
relevant cached selector from re-reselect area of the code:

```ts
const squareId = (_: StoreType, squareId: string, dateTimeString: string, tableId: number) =>
  squareId;
const dateTimeString = (_: StoreType, squareId: string, dateTimeString: string, tableId: number) =>
  dateTimeString;
const tableId = (_: StoreType, squareId: string, dateTimeString: string, tableId: number) =>
  tableId;

const getSquareById = (
  calendarSquaresByTableAndTime: AdditionalInventoryAndTablesWithSlots,
  squareId: string,
  dateTimeString: string,
  tableId: number
): NormalDetailedSquare | undefined => {
  const { tablesWithSlots } = calendarSquaresByTableAndTime;
  if (!tablesWithSlots) {
    return undefined;
  }
  const table = tablesWithSlots[tableId];
  if (!table) {
    return undefined;
  }
  const slot = table.slots;
  if (!slot) {
    return undefined;
  }
  const slotAtDateTime = slot[dateTimeString];
  if (!slotAtDateTime) {
    return undefined;
  }
  if (!slotAtDateTime.calendarSquaresById) {
    return undefined;
  }
  return slotAtDateTime.calendarSquaresById[squareId];
};

export const getCalendarSquareFromCalendar = createCachedSelector(
  calendarSquaresByTableAndTimeSelector,
  squareId,
  dateTimeString,
  tableId,
  getSquareById
)({
  keySelector: (_: StoreType, squareId: string, dateTimeString: string, tableId: number) =>
    squareId,
  cacheObject: new LruObjectCache({ cacheSize: 1000 }),
});
```

**Comment by kelly-tock** [comment]
getting cache hit right away with 4.0.0, but not with 4.1.5

**Comment by kelly-tock** [comment]
oh and the selector call in the function component: 

```tsx
const calendarSquare = useSelector(function calendarSquareUseSelector(state: StoreType) {
    return getCalendarSquareFromCalendar(state, calendarSquareId, dateTimeString, tableId);
  });

```

**Comment by markerikson** [comment]
Most of that looks like `re-reselect` internals, tbh.  

Does any of this happen with just `reselect` by itself, or only with `re-reselect`?

**Comment by kelly-tock** [comment]
The internals in the 4.1.5 are reselect internals, will screenshot those shortly. will profile with just reselect way as well.

**Comment by markerikson** [comment]
@kelly-tock fwiw the thing that would help the most is a sandbox that specifically replicates what you're seeing here.

**Comment by kelly-tock** [comment]
totally get it. so far I have a decent sandbox, but am unable to reproduce. the app i'm working on is incredibly complex(I know we all say that), so it has been hard to diagnose thus far.

**Comment by kelly-tock** [comment]
my guess is that is a huge number of small things adding up.

**Comment by kelly-tock** [comment]
every `memoize` in that stack for 4.1.5 is coming from this block in reselect in defaultMemoize:

![image](https://user-images.githubusercontent.com/39061683/146599533-38c097ab-3610-468b-8610-d2e9947fcd42.png)

every anonymous there in the long stack of back and forth between anonymous and memoize is in this part of reselect:

![image](https://user-images.githubusercontent.com/39061683/146599688-7bcd12aa-7ec0-4d1b-af4c-e5c22032c1ed.png)


in the 4.0.0 version it stops at re-reselect, guessing because its a cache hit, but I can't see why it wouldn't be a cache hit in both cases.

**Comment by kelly-tock** [comment]
with just regular reselect selector, same result for 4.1.5 and re-reselect:
```ts
export const getCalendarSquareCalendar = createSelector(
  calendarSquaresByTableAndTimeSelector,
  calendarSquareId,
  dateTimeString,
  tableId,
  getCalendarSquareById
);
```

and 
![image](https://user-images.githubusercontent.com/39061683/146800646-4c2cc49f-e95e-415a-bb4d-d501abcb3023.png)

**Comment by kelly-tock** [comment]
looks the same with a max size option as well:

```ts
export const getCalendarSquareCalendar = createSelector(
  calendarSquaresByTableAndTimeSelector,
  calendarSquareId,
  dateTimeString,
  tableId,
  getCalendarSquareById,
  {
    memoizeOptions: {
      maxSize: 1000,
    },
  }
);
```

**Comment by kelly-tock** [comment]
also tried above with the result option, using shallowEqual. same result. tried a pattern for creating a new selector instance for each component instead, similar results.

**Comment by markerikson** [comment]
@kelly-tock Can you put together a CodeSandbox or repo that demonstrates this? I really need something I can try to dig into and profile myself.

**Comment by kelly-tock** [comment]
yeah, working on it. just wanted to try out the suggestions above, and report back. here's 4.0.0 with creating a new selector instance for each component:

![image](https://user-images.githubusercontent.com/39061683/146806146-66115922-1e5c-4a5c-b5f8-890fc5530000.png)

**Comment by kelly-tock** [comment]
and runs just night and day faster.

**Comment by markerikson** [comment]
I'd really like to look at this sometime in the next few days. Could you please put together a CodeSandbox or a Github repo that fully demonstrates the issue happening?  That would really help me be able to investigate what's going on.

**Comment by kelly-tock** [comment]
I’ve been on vacation. I’ll link what I have so far, but if there’s a way to send a private profile file from chrome we can maybe do that? Our app is incredibly complex with 100s of selectors building different views of data

**Comment by markerikson** [comment]
You could zip it and send DM it to me over on Discord, but I really am going to need to actually dig into real running code to see what's going on myself.  Doesn't have to be the full app or anything - a minimal repro would be wonderful.

**Comment by kelly-tock** [comment]
I haven't gotten it fully repro-ing, but heres what I have so far:
https://github.com/kelly-tock/reselect-repro

What we do a lot of currently is receive things as arrays, and re-map them into object lookups by id. 

we also compute various properties from combining several data sources, and have thousands of items. 

when we use 4.0.0 performance is fine, but 4.1.5 makes the app completely unusable.

**Comment by kelly-tock** [comment]
let me know if you would like the profiles, as recreating this so far has been problematic.

**Comment by markerikson** [comment]
Thanks.  Looks like you've got a full-ish app there - can you point to the specific relevant bits or summarize the steps to reproduce the issue?

And yeah, if you can attach the perf traces (zipped) that would probably be helpful too.

Most likely won't have time to look at this until later this week.

**Comment by markerikson** [comment]
@kelly-tock : hey, if you're having trouble recreating the issue standalone, could you try using https://replay.io to record a replay of the bug on your own system? I haven't used it for real yet, but it ought to be tailor-made for this kind of a bug investigation scenario.

**Comment by kelly-tock** [comment]
will send profiles and try this out today. this is a fun one!

**Comment by kelly-tock** [comment]
@markerikson what discord are you referring to?  reactiflux?  I can dm you about the replay.io and profile stuff.

**Comment by kelly-tock** [comment]
can't post that link here, would need your email address to add to it if you want to dm me there.

**Comment by markerikson** [comment]
FYI I tried to do some analysis on this today based on the perf profiles, as well as a Replay of the demo app.  No concrete answers yet - I'll have to poke at it further.

I actually did the investigation as part of a livestream - that video is up on Youtube if you want to see my process:

https://www.youtube.com/watch?v=GvLGEuxV6BU

**Comment by kelly-tock** [comment]
I believe we do have a selector with that in our codebase that I had forgotten about. Will dig that up as well.  Watching the process is super interesting thanks for doing that!

**Comment by kelly-tock** [comment]
Also, in dev tools you can hit command + f/control+f and find the function. That is why I named the function in the useAppSelector call.

**Comment by kelly-tock** [comment]
a handful of selectors are created using the method in the docs with lodash.isEqual:

https://github.com/reduxjs/reselect#customize-equalitycheck-for-defaultmemoize

```
import { createSelectorCreator, defaultMemoize } from 'reselect';
import _ from 'lodash';

export const createDeepEqualSelector = createSelectorCreator(defaultMemoize, _.isEqual);
```

**Comment by kelly-tock** [comment]
ok, interesting, as an experiment I changed it to be this:
```
import { createSelectorCreator, defaultMemoize } from 'reselect';
// import _ from 'lodash';

export const createDeepEqualSelector = createSelectorCreator(defaultMemoize);
```

i'm guessing this is the same as just the normal `createSelector` with this?  

and this caused the app to behave normally again. maybe this is a clue as to where the issue is?

**Comment by koenoe** [comment]
> ok, interesting, as an experiment I changed it to be this:
> 
> ```
> import { createSelectorCreator, defaultMemoize } from 'reselect';
> // import _ from 'lodash';
> 
> export const createDeepEqualSelector = createSelectorCreator(defaultMemoize);
> ```
> 
> i'm guessing this is the same as just the normal `createSelector` with this?
> 
> and this caused the app to behave normally again. maybe this is a clue as to where the issue is?

I've had some issues with re-reselect and reselect (latest version) too. Decided to drop re-reselect and create my own simple implementation on top of the new API of reselect:
```
import { createSelectorCreator, defaultMemoize, EqualityFn } from 'reselect';

export const createCachedSelector = (equalFn: EqualityFn) =>
  createSelectorCreator(defaultMemoize, {
    maxSize: 100,
    resultEqualityCheck: equalFn,
  });
```

Example usage:
```
export const selectMovie = createCachedSelector(
  (prev: Movie, next: Movie) => prev.id === next.id,
)(
  selectContext,
  (state: MoviesMachine, id: number) => id,
  (context, id) => context[id],
);
```

It's very simple and might not be as advanced as re-reselect, but it does the job for me.

**Comment by markerikson** [comment]
It's a few years later, and this has gone stale.  I assume the majority of the issue was use of `deepEqual`.  If anyone has further details, let us know.
